src/lib/homestore_backend/tests/homeobj_fixture.hpp

```diff
@@ -24,123 +25,243 @@ using homeobject::ShardError;
 using namespace homeobject;
 
 #define hex_bytes(buffer, len) fmt::format("{}", spdlog::to_hex((buffer), (buffer) + (len)))
+
+extern std::unique_ptr< test_common::HSReplTestHelper > g_helper;
+
 class HomeObjectFixture : public ::testing::Test {
 public:
-    std::shared_ptr< FixtureApp > app;
-    std::shared_ptr< homeobject::HomeObject > _obj_inst;
-    std::random_device rnd{};
-    std::default_random_engine rnd_engine{rnd()};
+    std::shared_ptr< homeobject::HSHomeObject > _obj_inst;
+
+    // Create blob size in range (1, 16kb) and user key in range (1, 1kb)
+    HomeObjectFixture() : rand_blob_size{1u, 16 * 1024}, rand_user_key_size{1u, 1024} {}
 
     void SetUp() override {
-        app = std::make_shared< FixtureApp >(true /* is_hybrid */);
-        _obj_inst = homeobject::init_homeobject(std::weak_ptr< homeobject::HomeObjectApplication >(app));
+        _obj_inst = std::dynamic_pointer_cast< HSHomeObject >(g_helper->build_new_homeobject());
+        g_helper->sync();
+    }
+
+    void TearDown() override {
+        g_helper->sync();
+        _obj_inst.reset();
+        g_helper->delete_homeobject();
+    }
+
+    void restart() {
+        g_helper->sync();
+        trigger_cp(true);
+        _obj_inst.reset();
+        _obj_inst = std::dynamic_pointer_cast< HSHomeObject >(g_helper->restart());
+        // wait for leader to be elected
+        std::this_thread::sleep_for(std::chrono::seconds(5));
     }
 
-    void TearDown() override { app->clean(); }
+    // schedule create_pg to replica_num
+    void create_pg(pg_id_t pg_id, uint32_t replica_num = 0) {
```

Comment:
why do we need create_pg with replica_num , isnt it executed only on leader and other replica's will get it via raft.

Reply:
HSHomeObject::_create_pg will first create a repl_dev, then replicate a create_pg message across the raft group(repl_dev). only until we have a repl_dev , we can have a leader.

so, in homeobject level, before we create a pg(HSHomeObject::_create_pg is called), we don`t have a repl_dev for this pg and thus we do not have a leader. here `replica_num` means which replica will be the leader when creating repl_dev of this pg